### DISCUSSION TITLE
Adding the React UI console for testing.

### DISCUSSION THREAD

**Discussion by Aditya-eddy** [keploy/keploy#1633] [Ideas]
Hey Keploy Community,
Creating a react UI console for testing will be a big plus for testing with keploy.
I've created a very basic react website. Which Will provide the onine testing.
1. Created a HomePage for website. (/test-generator)
![Screenshot from 2024-02-29 11-44-48](https://github.com/keploy/keploy/assets/70089590/489cfbd2-a748-46c3-a50b-d0aa7ac29d7b)
2. Created different Tests for the User according to language they want to perform in.
![Screenshot from 2024-02-29 11-45-56](https://github.com/keploy/keploy/assets/70089590/daad73f8-dbba-4bcb-8896-025c56e894eb)
3. I've also created a Editor (CodeMirror) and Terminal for the keploy. 
![Screenshot from 2024-02-29 11-46-07](https://github.com/keploy/keploy/assets/70089590/a8af4778-35c4-424d-b2f7-5bd6dafe93f9)
![Screenshot from 2024-02-29 11-46-03](https://github.com/keploy/keploy/assets/70089590/fce36291-1376-43f5-966e-db50c511b710)

Terminal takes the following command:-record,test,clear and if there is command entered by user it return not found.
Currently It has a very basic UI but will be working on it. 
I've also addded the Tutorial section which will tell the user about the commands and How he can perform the testing.
I've also kept it styling according with the blog-website but can also be changed.
Will be waiting on feedback from you guys. Thanks!!

**Comment by Aditya-eddy** [comment]
Hey Everyone. Today I tried to give it replit like feel. I think these approach will be more suitable with results that we want. 
1. Make editor and terminal side by side. Gave them both Monokai theme. (Can be change if we want a lighter theme). I Have also made them have some code depending on their language already. Both the console and editor are working.
Will working more on the UI.
One question I have is will this website move into direction of a dark theme or light theme Intitally? Will be waiting on feedback. Thanks!!
![Screenshot from 2024-03-01 15-43-16](https://github.com/keploy/keploy/assets/70089590/e7e80dd8-04d5-4724-8fc9-81078ba3dbea)
![Screenshot from 2024-03-01 15-43-08](https://github.com/keploy/keploy/assets/70089590/5c498985-ae02-480d-9b04-5943bbb13fb3)

![Screenshot from 2024-03-01 15-43-26](https://github.com/keploy/keploy/assets/70089590/b9ca6410-fa24-456a-b950-09233735acea)

**Comment by Aditya-eddy** [comment]
Hey everyone Today created a new UI for the cards displaying the Index. Where will be displaying all the tests Present. I'll attach the video with the comments. Please review when you get to it.
[Screencast from 05-03-24 04:14:29 PM IST.webm](https://github.com/keploy/keploy/assets/70089590/fea70d35-5750-487b-8e79-d0e0c6072eca)

**Comment by nehagup** [comment]
Hey @Aditya-eddy ! We're working internally to define the requirements for this project better. Please allow us a few days to get back to you on this.

**Comment by nehagup** [comment]
Hi @Aditya-eddy here's a very high-level wireframe/idea of how we're planning to place the components. Please note that after the discussions we've decided not to move forward with the separate code-sandbox development, we believe that can be done leveraging existing sandboxing tools. 


This project will majorly revolve around adding this Auto-Test generation demo/play area on the keploy website only. 

<img width="999" alt="Screenshot 2024-03-12 at 5 26 42 PM" src="https://github.com/keploy/keploy/assets/15074229/e69d59ce-606b-450e-b361-868ab0afa833">

While you can add the styling, animations and other vanity things later, it's advised to build a prototype for the above design and become aware of the scenarios that can happen in this playground.

**Reply by Aditya-eddy** [reply]
Okay @nehagup this seems exiciting.

**Comment by Aditya-eddy** [comment]
Hey Everyone, By taking  the functionalities listed above in consideration I will describe the features of the project given below.
(Screenshot attached Below)
1. **For the Side Tree:-**
I have taken the general medhot of making a side tree by dividing them in following format of rootDir-
`let rootDir: Directory = {
    id: "0",
    name: "root",
    parentId: undefined,
    type: Type.DIRECTORY,
    depth: 0,
    dirs: [],
    files: []
  };`
  Hence displaying the Side Tree won't be a problem. For fetching the data in the side tree I would recommend using socket or any low latency bidirectional communication here. It'll be helpful.
 
 2. **Code Editor:-**
  Code Editor is monaco code editor. Whenever we click anything on the side tree the content will be displayed in the code editor. Currently It supports one tab but that can be fixed for later.
  From Code Editor we can send the data to the backend by saving the file.

  3.**Terminal**
        I have taken the Terminal from the previous designed model. For sending commands and receiving the data.
        We can type the command in the terminal which will trigger the backend to do the Following process. Also socket will provide high speed. which will be using in fetching data.

4. **Content Area**:-
  I have not defined much here just wrote some dummy text. Progess Bar and Language button are still remaining.

 **Overview:-**
  First the user will select the language and based on that language some files will fetched and shown in the side-tree.Upon clicking on it user can edit the file. Edited file will be send to the backend and saved their. Then the user can run commands on the terminal triggering the backend to send logs,test etc.
  
**FeedBack**
Please guide me on this. Correct me anywhere I am wrong or where I can improve. Will waiting from feedback from you guys. Thanks.

ScreenShot:-
![Screenshot from 2024-03-14 16-46-37](https://github.com/keploy/keploy/assets/70089590/ef514bef-6be4-4db0-99a1-cdcf70e7d435)

cc:- @nehagup , @Hermione2408, @charankamarapu

**Comment by nehagup** [comment]
@Aditya-eddy Please submit the repo of your prototype for review and start working on the proposal.

**Comment by Aditya-eddy** [comment]
Hey everyone I am excited to announce that the react console can now capture test Cases. I'll attach the demonstration below.
What I have done is created a scrappy backend to give us a glimpse how will the react console would look like. Ofc many functionalities are still left and I have done this using express-mongo only.
It's a prototype which will help us in the future for online test-case generation.
I'll make the repo public soon.
[Screencast from 20-03-24 09:44:11 AM IST.webm](https://github.com/keploy/keploy/assets/70089590/77dd4c2e-bfe8-4724-9622-fbbb69853f5f)

Feedback would be appreciated from you guys.
cc - @nehagup , @Hermione2408 , @charankamarapu .

**Comment by charankamarapu** [comment]
I think the commands which are present in your demo like curl, fetch. I think either we have to provide exact commands like "keploy record -c "xxxx" " or may be provide a button each action and once you press that button that command will be ran in the terminal. I am suggesting this because users may get confused between the original commands like "keploy record -c "xxxx" " and UI specific commands which you have showed in the demo.

**Reply by Aditya-eddy** [reply]
Yes I have just  made these commands for ease. They can be changed easily! . For fetching and saving the data we can create a button for it.

**Reply by Hermione2408** [reply]
Hey instead of copy pasting the commands , I was also thinking the same as @charankamarapu  suggested to use record,replay buttons that will directly execute the commands, that would bring the number of user actions required to execute a command from 3 to 1.

**Reply by Aditya-eddy** [reply]
Okay I'll add the original commands and specific buttons for fetching the data ,running the keploy and testing with keploy.

**Comment by Aditya-eddy** [comment]
Hey @charankamarapu , @Hermione2408 taking the feedback from you guys. Add a task bar. First I'll attach the video for recording.
1. Here I have demonstrated the recording of the task cases. Added the task bar with record,test,stop,fetch. (ps-Sorry for laggy Video).
[good.webm](https://github.com/keploy/keploy/assets/70089590/6f053be6-433c-45c9-99dd-d0ebae5201a1)
2. Here I have attached the screenshots when the testsing is started in the application. It will create a log file in the directory and percentage fo the testing with jest and combined converage has been stated here. 
![Screenshot from 2024-03-22 14-13-18](https://github.com/keploy/keploy/assets/70089590/d5437bdd-ede6-456e-94b4-248c427ee591)
![Screenshot from 2024-03-22 14-13-07](https://github.com/keploy/keploy/assets/70089590/8ad1a1f6-1731-40b6-b656-7d2e5b9a4783)
![Screenshot from 2024-03-22 14-10-55](https://github.com/keploy/keploy/assets/70089590/c9167c41-7976-4350-b920-c344316893db)

- Next I am thinking of  the keploy record test -c "xxxx" functionality and add a graph for combined test Coverage in the Side content
- Then I would probably extent it to other application Go-test, J-unit,Py-Test.

Feedback would be Appreciated from you guys.

**Comment by Aditya-eddy** [comment]
Hey Mentors I have the published the repo [here](https://github.com/Aditya-eddy/prototype?tab=readme-ov-file).
Please give your feedback on it.
cc- @nehagup , @Hermione2408 , @charankamarapu

**Comment by nehagup** [comment]
@Aditya-eddy can you please submit the code for review? You can share it in a personal repo for now.

**Reply by Aditya-eddy** [reply]
Here is the Protoype's [link](https://github.com/Aditya-eddy/prototype?tab=readme-ov-file).
